### ISSUE TITLE
Change the way spacy works - Custom properties for training and prediction

### ISSUE THREAD

**Issue by ghost** [explosion/spaCy#4864]
Hi,
I originally posted this question on stackexchange but haven't got any answer yet.
[change-the-way-spacy-works-custom-properties-for-training-and-prediction](https://datascience.stackexchange.com/questions/65477/change-the-way-spacy-works-custom-properties-for-training-and-prediction)
Can someone please suggest on this:

Spacy detects the entities using its predefined algorithm. It parses tokens in text considering position of tokens with respect to tokens surrounding it. It also takes into consideration the POS tagging for these tokens.

However, I believe it misses the position of tokens above and below (For example in a tabular data) or it also misses few properties of text like if it is underline etc. This statement is based on my understanding. Please correct me if I am wrong in these.

Now the question is, **can such properties be taken into consideration while doing training and prediction of entities**? I have seen Extension Attributes, but these do not play role during training and prediction but work as meta data.

**Comment by adrianeboyd** [comment]
Your understanding is correct, except that as of v2.0 the tagger / parser / NER models are all independent of each other so the NER model doesn't take output from the tagger or parser into account.

Unfortunately spacy's models and features are not easy to extend currently. Work is in progress in this area since it's becoming a more frequent request from users, so I hope we can provide a more helpful response in the future.

**Comment by svlandeg** [comment]
Happy to announce that we just released [v3.0.0rc1](https://github.com/explosion/spaCy/releases/tag/v3.0.0rc1)!

With v3, you'll be able to extend/change the model implementations much more easily. You can find the details of the default model implementations from the core library here: https://nightly.spacy.io/api/architectures. One that could be of particular interest to you is `MultiHashEmbed` that allows you to define a list of attributes.

You can also [implement](https://nightly.spacy.io/usage/layers-architectures) your own model now, [register](https://nightly.spacy.io/api/top-level#registry) it, and then use it in your [config](https://nightly.spacy.io/usage/training#config) file.

So, long story short: this feature request is probably too specific to be included in the main library, but we do hope that the revamped v3 makes it much more easy to implement and customize this, tailored to your specific needs. If you end up developing a package that you think would be useful to others as well, please also consider adding it to the [spaCy universe](https://nightly.spacy.io/universe) :-)

I'll close this in the meantime - but feel free to open a new one if you run into issues!
